HP-1099: Render the verify email notification only with certain amr

NikoHelle opened the pull request:
First commit includes missing tests for useProfile. Tests were added because second commit includes a change how useProfile handles the profile returned from Tunnistamo. According to the OIDC spec, Profile.amr should be an array of strings, but it is not. Now it is converted to an array, if it is not.

EmailEditor should show the "verify your email"-notification only if user's amr is 'heltunnistussuomifi'. Currently that is the only case, so the check is simple. In other cases the backend won't send an email and the notification should not be shown.

NikoHelle committed:
(HP-1099) Render the verify email notification only with certain amr

The notification should only be shown if user's amr is 'heltunnistussuomifi'.

In other cases the backend won't send an email and the notification should not be shown.

Because useProfile is mocked AND imported, the tunnistusSuomifiAMR cannot be reused inside this test.

Typescript compiler will fail. That's why 'tunnistusSuomifi' from AmrStatic is used and 'tunnistusSuomifiAMR' is re-declared.

akikoskinen commented on src/profile/components/profileInformation/profileInformationAccountManagementLinkUtils.ts:
```diff
@@ -52,3 +52,7 @@ export function getAmrUrl(authenticationMethodReference: AMRStatic): string {
 }
+
+export function hasTunnistusSuomiFiAmr(profile: Profile | null): boolean {
+  return profile?.amr === tunnistusSuomifiAMR;
```
According to [the OIDC specification](https://openid.net/specs/openid-connect-core-1_0.html#IDToken), the `amr` claim is an **array of strings**. It could be that Tunnistamo erroneously provides a single string. I think both styles should be handled here, a single string, and an array of strings.

NikoHelle replied:
Very good point, I didn't know Tunnistamo is not compliant with the specs. I checked that Helsinki Tunnistus (Keycloak) returns an array in amr. But Profiili does not use Tunnistus (yet).

Tunnistamo never returns an array, so the possibility should not be added to Profiili at the moment. The code would be redundant. It is clear this must be changed when Tunnistamo won't be used anymore.

akikoskinen replied:
That would be knowingly leaving a bug in the code waiting to bite. I wouldn't leave it there. Who remembers to fix it on time before it bites?

NikoHelle replied:
From Profile's point of view the bug is in Tunnistamo. :)

I'll change this to be future proof.

Maybe report a bug for Tunnistamo?

akikoskinen replied:
> ...the bug is in Tunnistamo. :)
> Maybe report a bug for Tunnistamo?

Yep, that's known. Unfortunately it's difficult to change because it would break backward compatibility :disappointed:

NikoHelle replied:
Tests added and amr is now converted to "string[]".

akikoskinen reviewed (changes requested):
Should the `amr` values (plural, as I think there could be more than one of them) that trigger showing the notification, be configurable? I suppose that the set of those values can change and also be specific to different environments.

NikoHelle commented:
> Should the `amr` values (plural, as I think there could be more than one of them) that trigger showing the notification, be configurable? I suppose that the set of those values can change and also be specific to different environments.

I added just this simple check, because it is unclear what kind of scenarios there will be in the future. According to Slack-conversations nobody knows exactly how the email verification will evolve and when those are sent in the future.

So until there is a second scenario, which might show the way, I'd just stick to this simple way. If the email sending depends only on amr and that changes in different environments, then it's definitely a good idea to add it to env config.

NikoHelle committed:
(HP-1099) Added tests for useProfile.ts hook

Created mock data creator for OIDC User and User.Profile

To use them with jest.mock the function must start with "mock"

Added mock data creators to older tests

NikoHelle committed:
(HP-1099) Force amr to be an array

Tunnistamo returns amr as a string, but OIDC spec states it is an array of strings

Amr is now converted to an array if it is a string. Undefined or empty amr is converted to an empty array

Removed the AMR type as it was useless. typeof config.helsinkiAccountAM is "any", so the AMR type is then also "any".

akikoskinen commented on src/auth/__tests__/useProfile.test.tsx:
```diff
@@ -0,0 +1,134 @@
+    expect(getProfile()).toEqual(noProfile);
+  });
+
+  it('should return an empty array is arm is undefined', async () => {
```
First "is" should probably be "if".

NikoHelle replied:
Done!

akikoskinen commented on src/auth/__tests__/useProfile.test.tsx:
```diff
@@ -0,0 +1,134 @@
+    expect(getProfile()).toEqual(profileWithConvertedAmr);
+  });
+
+  it('should return null if user.expired !== false', async () => {
```
This test name could be more readable. Perhaps something like:
> should provide no profile if it has expired

NikoHelle replied:
The returned value of user.expired can be "undefined", if the expires_in the token is "undefined".  That is why "user.expired !== false" is used instead of just "!user.expired". 

But the name of the test does not need to reflect that. Added another test for undefined.

akikoskinen commented on src/profile/components/emailEditor/__tests__/EmailEditor.test.tsx:
```diff
@@ -27,6 +27,19 @@ import { EditDataType, EmailValue } from '../../../helpers/editData';
+import { mockProfileCreator } from '../../../../common/test/userMocking';
+import { AMRStatic } from '../../../../auth/useProfile';
+
+type TestAmr = Extract<AMRStatic, 'google'> | 'foobar';
```
Instead of "foobar", could the test amr be something more descriptive, like perhaps "tunnistus-suomi-fi"?

NikoHelle replied:
This is intentionally silly value so it is not mistaken to the actual value.

akikoskinen replied:
It could still have some meaning. Perhaps something like "fake-suomi-fi-amr-for-test"...

NikoHelle replied:
Switched to a value defined in "AmrStatic", because it is most convenient. Switching to a value longer than 10 chars will trigger linting error "sonarjs/no-duplicate-string"

akikoskinen commented on src/profile/components/profileInformation/profileInformationAccountManagementLinkUtils.ts:
```diff
@@ -52,3 +56,7 @@ export function getAmrUrl(authenticationMethodReference: AMRStatic): string {
 }
+
+export function hasTunnistusSuomiFiAmr(profile: Profile | null): boolean {
+  return extractSingleAmr(profile) === tunnistusSuomifiAMR;
```
This should check if `tunnistusSuomifiAMR` is present _anywhere_ in the `profile`'s `amr` array.

NikoHelle replied:
When does the amr array has length !== 1? The spec says it is possible, but in our environments?

From the amr spec: "Parties using this claim will need to agree upon the meanings of the values used, which may be context-specific." Is there an agreement somewhere?

UI does not process more than one amr value. For example the ProfileInformation component renders only one ProfileInformationAccountManagementLink. If there are no amr values, the component is not rendered so user won't notice it.

akikoskinen replied:
Currently I guess there's only one `amr` value. I'm not aware of any case where there would be more than one value in the list. I just thought that finding the correct value from anywhere in the array would require about the same amount of code, so why not do it.

NikoHelle replied:
Number of items in the "amr" array affects other places too. Now it can be assumed there is 0 or 1 item and the value can just be picked. Otherwise all checks should search if the value exists anywhere in the array.

akikoskinen replied:
Got it. Since that's the case, I guess the multi-value support can be left for another time.

akikoskinen commented on src/profile/components/profileInformation/profileInformationAccountManagementLinkUtils.ts:
```diff
@@ -5,8 +5,12 @@ import {
 import config from '../../../config';
 
+function extractSingleAmr(profile: Profile | null): string {
+  return profile ? profile.amr[0] : '';
```
What if the `amr` array is empty? Should a check be made before executing `profile.amr[0]`?

NikoHelle replied:
This depends on the conversation above.

akikoskinen replied:
This is also relevant if the `amr` claim can be missing or an empty string (in the case of Tunnistamo, which doesn't provide the amrs in an array but as a string).

NikoHelle replied:
The string value is converted to an array of single string. But anyway the return value of that function requires "| undefined"

akikoskinen reviewed (commented):
Some smallish comments.
